localstack/services/s3/presigned_url.py

```diff
@@ -0,0 +1,570 @@
+    """
+    Pre-signed URL with PUT method (typically object upload) should return an empty body
+    """
+    if (
+        not context.request.method == "PUT"
+        or not is_presigned_url_request(context)
+        or response.status_code >= 400
+    ):
+        return
+    else:
+        response.data = b""
+
+
+def s3_presigned_url_request_handler(_: HandlerChain, context: RequestContext, __: Response):
+    """
+    Handler to validate S3 presigned URL. Checks the validity of the request signature, and raises an error if
+    `S3_SKIP_SIGNATURE_VALIDATION` is set to False
+    """
+    if context.service.service_name != "s3":
+        return
+
+    if not is_presigned_url_request(context):
+        # validate headers, as some can raise ValueError in Moto
+        _validate_headers_for_moto(context.request.headers)
+        return
+    # will raise exceptions if the url is not valid, except if S3_SKIP_SIGNATURE_VALIDATION is True
+    # will still try to validate it and log if there's an error
+
+    is_v2 = all(p in context.request.args for p in SIGNATURE_V2_PARAMS)
+    is_v4 = all(p in context.request.args for p in SIGNATURE_V4_PARAMS)
+
+    if not is_v2 and any(p in context.request.args for p in SIGNATURE_V2_PARAMS):
+        LOGGER.info("Presign signature calculation failed")
+        ex: AccessDenied = create_access_denied_missing_parameters_sig_v2()
+        raise ex
+
+    elif is_v2 and not is_v4:
+        validate_presigned_url_s3(context)
+
+    if not is_v4 and any(p in context.request.args for p in SIGNATURE_V4_PARAMS):
+        LOGGER.info("Presign signature calculation failed")
+        ex: AuthorizationQueryParametersError = create_authorization_query_parameters_error()
+        raise ex
+
+    elif is_v4 and not is_v2:
+        validate_presigned_url_s3v4(context)
+
+    _validate_headers_for_moto(context.request.headers)
+    LOGGER.debug("Valid presign url.")
+
+
+def is_expired(expiry_datetime: datetime.datetime):
+    now_datetime = datetime.datetime.now(tz=expiry_datetime.tzinfo)
+    return now_datetime > expiry_datetime
+
+
+def is_presigned_url_request(context: RequestContext) -> bool:
+    """
+    Detects pre-signed URL from query string parameters
+    :param context: the request context from the handler chain
+    """
+    # Detecting pre-sign url and checking signature
+    query_parameters = context.request.args
+    return any(p in query_parameters for p in SIGNATURE_V2_PARAMS) or any(
+        p in query_parameters for p in SIGNATURE_V4_PARAMS
+    )
+
+
+def _get_aws_request_headers(werkzeug_headers: Headers) -> HTTPHeaders:
+    """
+    Converts Werkzeug headers into HTTPHeaders() needed to form an AWSRequest
+    :param werkzeug_headers: Werkzeug request headers
+    :return: headers in HTTPHeaders format
+    """
+    headers = HTTPHeaders()
+    for key, value in werkzeug_headers.items():
+        headers[key] = value
```

Comment:
Good catch! I'll check exactly how `HTTPHeaders` return multiple values and deal with it in consequence

Reply:
HTTPHeaders are actually a subclass from `email.message.Message`.
See https://docs.python.org/3/library/http.client.html#httpmessage-objects and https://docs.python.org/3/library/email.compat32-message.html#email.message.Message.__setitem__
It will append the values together when called multiple times on the same key, so the right behaviour out of the box. I had a hunch when they deleted the value before setting it in `botocore`. But now it's confirmed, thanks! I added the link to the doc from Python in the function.